expr/cutter.go

```diff
@@ -1,199 +1,146 @@
-		if s.Equal(el) {
-			return true
-		}
+	if rec != nil {
+		c.dirty = true
 	}
-	return false
+	return rec, nil
 }
 
-// setBuilder creates a builder for the regular form of cut, where a
-// set of fields are to be cut from a record and passed on.
-//
-// Note that unlike for the complement form, we don't strictly need a
-// different columnbuilder or set of resolvers per input type
-// here. (We do need a different outType). Since the number of
-// different input types is small wrt the number of input records, the
-// optimization consisting of having a single columnbuilder and
-// resolver set doesn't seem worth the added special casing.
-func (c *Cutter) setBuilder(r *zng.Record) (*cutBuilder, error) {
-	// Build up the output type.
-	var fields []field.Static
-	var resolvers []Evaluator
-	var outColTypes []zng.Type
-	for i, resolver := range c.resolvers {
-		val, err := resolver.Eval(r)
-		if err != nil || val.Type == nil {
-			// The field is absent, so for this input type, ...
-			if c.strict {
-				// ...produce no output.
-				return nil, nil
-			}
-			// ...omit the field from the output.
-			continue
+func (c *Cutter) lookupTypeRecord(types []zng.Type) (*zng.TypeRecord, error) {
+	id := c.outTypes.Lookup(types)
+	typ, ok := c.recordTypes[id]
+	if !ok {
+		cols := c.builder.TypedColumns(types)
+		var err error
+		typ, err = c.zctx.LookupTypeRecord(cols)
+		if err != nil {
+			return nil, err
 		}
-		fields = append(fields, c.fields[i])
-		resolvers = append(resolvers, resolver)
-		outColTypes = append(outColTypes, val.Type)
+		c.recordTypes[id] = typ
 	}
-	if len(fields) == 0 {
-		return nil, nil
-	}
-	builder, err := builder.NewColumnBuilder(c.zctx, fields)
-	if err != nil {
-		return nil, err
-	}
-	cols := builder.TypedColumns(outColTypes)
-	outType, err := c.zctx.LookupTypeRecord(cols)
-	if err != nil {
-		return nil, err
-	}
-	return &cutBuilder{resolvers, builder, outType}, nil
+	return typ, nil
 }
 
-func (c *Cutter) builder(r *zng.Record) (*cutBuilder, error) {
-	if c.complement {
-		return c.complementBuilder(r)
+func fieldList(fields []Evaluator) string {
+	var each []string
+	for _, fieldExpr := range fields {
+		s := "<not a field>"
+		if f, err := DotExprToField(fieldExpr); err == nil {
+			s = f.String()
+		}
+		each = append(each, s)
 	}
-	return c.setBuilder(r)
+	return strings.Join(each, ",")
 }
 
-// Cut returns a new record comprising fields copied from in according to the
-// receiver's configuration.  If the resulting record would be empty, Cut
-// returns nil.
-func (c *Cutter) Cut(in *zng.Record) (*zng.Record, error) {
-	cb, ok := c.cutBuilders[in.Type.ID()]
-	if !ok {
-		var err error
-		cb, err = c.builder(in)
-		if err != nil {
-			return nil, err
-		}
-		c.cutBuilders[in.Type.ID()] = cb
+func (c *Cutter) Warning() string {
+	if c.FoundCut() {
+		return ""
 	}
-	if cb == nil {
-		return nil, nil
+	list := fieldList(c.fieldExprs)
+	cmd := "pick"
+	if c.droppers != nil {
+		cmd = "cut"
 	}
```

Comment:
This part feels dicey.

How about a wrapper in compiler/proc.go like
```go
type prefixedWarningFunction {
	proc.Function
	prefix string
}

func (p *prefixedWarningFunction) Warning() string {
	if s := p.Function.Warning(); s != "" {
		return fmt.Sprintf("%s: %s", prefix, s)
	}
}
```
so that compileProc can say
```go
		return proc.NewApplier(pctx, parent, prefixedWarningFunction{cutter, "cut"}), nil
```
and
```go
		return proc.NewApplier(pctx, parent, prefixedWarningFunction{cutter, "pick"}), nil
```
?

Reply:
This is a less bad hack.  I added a warning prefix to proc.applier and created issue #1776.